src/helper.cc

```diff
@@ -573,6 +569,16 @@ Helper::Client::submit(const char * const buf, HLPCB * const callback, void *dat
     debugs(84, DBG_DATA, Raw("buf", buf, strlen(buf)));
 }
 
+void
+Helper::Client::callBack(Xaction &r)
+{
+    const auto callback = r.request.callback;
```

Comment:
Old code did not reset request.callback in some cases and, hence, could, in the presence of certain other bugs, call this callback twice without crashing. We now always reset that data field, so those bugs (if any) may cause a crash in our code. Let's check our preconditions:

```suggestion
    const auto callback = r.request.callback;
    Assure(callback);

```

Reply:
Ok, but  we should be able to call nil callback because r.request.data will be nil then.

Reply:
> Ok, but we should be able to call nil callback because r.request.data will be nil then.

Did you mean to say "should not be able"?

Reply:
Yes, I meant that :)

Reply:
> we should not be able to call nil callback because r.request.data will be nil then

Yes, r.request.data will indeed be nil during the second call, but nil r.request.data does not make cbdataReferenceValidDone() false because cbdataReferenceValidDone(x,y) effectively returns cbdataReferenceValid(x). The latter is true for nil x:

```C++
int
cbdataReferenceValid(const void *p)
{
    if (p == nullptr)
        return 1;       /* A NULL pointer cannot become invalid */
    ...
}
```